app/src/main/java/org/wikipedia/analytics/metricsplatform/ArticleEvent.kt

```diff
@@ -0,0 +1,316 @@
+                "wiki_db" to (fragment.title?.wikiSite?.dbName() ?: ""),
+                "version" to PROD_LINK_PREVIEW_VERSION
+            )
+            submitEvent(
+                "android.metrics_platform.article_link_preview_interaction",
+                getPageData(fragment),
+                customData
+            )
+        }
+    }
+
+    inner class ArticleLinkPreviewDialogInteraction(
+        private val wikiDb: String,
+        private val pageId: Int,
+        private val source: Int
+    ) {
+        private val PROD_LINK_PREVIEW_VERSION = 3
+        fun logLinkClick() {
+            submitEvent("linkclick")
+        }
+
+        fun logNavigate() {
+            submitEvent(if (Prefs.isLinkPreviewEnabled) "navigate" else "disabled")
+        }
+
+        fun logCancel() {
+            submitEvent("cancel")
+        }
+
+        private fun submitEvent(action: String) {
+            val customData = mapOf(
+                "action" to action,
+                "source" to source,
+                "time_spent_ms" to timer.elapsedMillis,
+                "wiki_db" to wikiDb,
+                "page_id" to pageId,
+                "version" to PROD_LINK_PREVIEW_VERSION
+            )
+            submitEvent(
+                "android.metrics_platform.article_link_preview_interaction",
+                null,
+                customData
+            )
+        }
+    }
+
+    /**
+     * Submits events to the Metrics Platform.
+     */
+    private fun submitEvent(
+        eventName: String,
+        pageData: PageData?,
+        customData: Map<String, Any>
+    ) {
+        MetricsPlatform.client.submitMetricsEvent(eventName, pageData, mergeData(customData))
+    }
+
+    private fun getPageData(fragment: PageFragment): PageData {
+        val pageProperties = fragment.page?.pageProperties
+        return PageData(
+            pageProperties?.pageId ?: 0,
+            pageProperties?.displayTitle ?: "",
+            pageProperties?.namespace?.code() ?: 0,
+            Namespace.of(pageProperties?.namespace?.code()!!).toString() ?: "",
+            pageProperties.revisionId.toInt() ?: 0,
+            pageProperties.wikiBaseItem ?: "",
+            fragment.model.title?.wikiSite?.languageCode ?: "",
+            null,
+            null,
+            null
+        )
+    }
+
+    /**
+     * Merges custom data with additional client metadata.
+     */
+    private fun mergeData(data: Map<String, Any>): Map<String, Any> {
+        return data + ApplicationData.getApplicationData()
+    }
+
+    /**
+     * Creates a new read-only map by replacing or adding entries to this map from another map.
+     *
+     * The returned map preserves the entry iteration order of the original map.
+     * Those entries of another map that are missing in this map are iterated
+     * in the end in the order of that map.
+     */
+    public operator fun <K, V> Map<out K, V>.plus(map: Map<out K, V>): Map<K, V> =
```

Comment:
The plus operator for two maps already exists in the standard library. We don't need to define it ourselves.

Reply:
gtk! removed